### ISSUE TITLE
Registration failed on sample PoC Java Application.

### ISSUE THREAD

**Issue by Kojiro-Goto** [StrongKey/fido2#18]
**Describe the bug**
Sign Up from PoC application doesn't finish successfully.

**To Reproduce**
Install FIDO2 Server on CentOS7(Linux ******* 3.10.0-1062.el7.x86_64 #1 SMP Wed Aug 7 18:08:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux) successfully.
Install PoC Java application successfully.
Change the value of keystore as follows.
1. java -jar keymanager.jar addaccesskey 〜
2. java -jar keymanager.jar deleteaccesskey　〜 <defaultaccesskey>
3. Change Access key and Secret key in the /applications/poc/WEB-INF/classes/resources/poc-configuration.properties file.
   * poc.cfg.property.accesskey and poc.cfg.property.secretkey
   * regeneratesigningkey doesn't work correctly. didn't update default key and create new one.
4. Restart glassfishd
5. Access to the top url of PoC.
6. Choose "Sign Up" and enter email address 
7. Receive invitation email and open the link in this email
8. Input infos and click "Sign Up" button.

Then the webapp shows error "An unexpected error occurred. Please check server logs for details.".

**Expected behavior**
Sign up finish successfully as same as DEMO(https://fido2.strongkey.com/)

**Server logs(domain1/logs/server.log)**
[2019-10-15T17:59:37.944+0900] [Payara 4.1] [INFO] [] [] [tid: _ThreadID=34 _ThreadName=http-thread-pool::http-listener-2(10)] [timeMillis: 1571129977944] [levelValue: 800] [[
  POST https://fido2test.ip3.co.jp:8181/api/domains/1/fidokeys/registration/challenge HTTP/1.1]]

[2019-10-15T17:59:38.716+0900] [Payara 4.1] [INFO] [CRYPTO-MSG-1052] [CRYPTO] [tid: _ThreadID=26 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1571129978716] [levelValue: 800] [[
  CRYPTO-MSG-1052: Default configuration properties:
        crypto.cfg.maxlen.1024charstring: 1024
        crypto.cfg.maxlen.10charstring: 10
        crypto.cfg.maxlen.128charstring: 128
        crypto.cfg.maxlen.16384charstring: 16384
        crypto.cfg.maxlen.16charstring: 16
        crypto.cfg.maxlen.2048charstring: 2048
        crypto.cfg.maxlen.2080charstring: 2080
        crypto.cfg.maxlen.20charstring: 20
        crypto.cfg.maxlen.256charstring: 256
        crypto.cfg.maxlen.32768charstring: 32768
        crypto.cfg.maxlen.32charstring: 32
        crypto.cfg.maxlen.4096charstring: 4096
        crypto.cfg.maxlen.4charstring: 4
        crypto.cfg.maxlen.512charstring: 512
        crypto.cfg.maxlen.5charstring: 5
        crypto.cfg.maxlen.64charstring: 64
        crypto.cfg.maxlen.65535charstring: 65535
        crypto.cfg.maxlen.6charstring: 6
        crypto.cfg.maxlen.8192charstring: 8192
        crypto.cfg.maxlen.8charstring: 8
        crypto.cfg.maxlen.9charstring: 9
        crypto.cfg.property.certalgorithm: EC
        crypto.cfg.property.certkeysize: 256
        crypto.cfg.property.certsigningalgorithm: SHA256withECDSA
        crypto.cfg.property.certvalidity: 3650
        crypto.cfg.property.cryptohome: /usr/local/strongkey/crypto
        crypto.cfg.property.cryptomodule.type: tpm
        crypto.cfg.property.cryptomodule.vendor: stm
        crypto.cfg.property.encsuffix: -ENC
        crypto.cfg.property.enctransform: AES/GCM/NoPadding
        crypto.cfg.property.fipsdrbg.personalizationstring:
        crypto.cfg.property.fipsdrbg.securitystrength: 256
        crypto.cfg.property.fipsmode: true
        crypto.cfg.property.hsm.certalgorithm: EC
        crypto.cfg.property.hsm.certkeycurve: NIST-P521
        crypto.cfg.property.hsm.certkeysize: 256
        crypto.cfg.property.hsm.certsigningalgorithm: SHA256withECDSA
        crypto.cfg.property.signing.ec.signaturealgorithm: SHA256withECDSA
        crypto.cfg.property.signing.keystorelocation: /usr/local/strongkey/skfs/keystores/signingkeystore.bcfks
        crypto.cfg.property.signing.rsa.signaturealgorithm: SHA256withRSA
        crypto.cfg.property.signing.truststorelocation: /usr/local/strongkey/skfs/keystores/signingtruststore.bcfks
        crypto.cfg.property.signsuffix: -SIGN
        crypto.cfg.property.xmlsignsuffix: -XML-SIGN]]

[2019-10-15T17:59:38.716+0900] [Payara 4.1] [INFO] [CRYPTO-MSG-1053] [CRYPTO] [tid: _ThreadID=26 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1571129978716] [levelValue: 800] [[
  CRYPTO-MSG-1053: Configuration log message: CRYPTO_HOME is: /usr/local/strongkey/crypto]]

[2019-10-15T17:59:38.716+0900] [Payara 4.1] [INFO] [CRYPTO-MSG-1053] [CRYPTO] [tid: _ThreadID=26 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1571129978716] [levelValue: 800] [[
  CRYPTO-MSG-1053: Configuration log message: Using crypto-configuration.properties from CRYPTO_HOME directory: /usr/local/strongkey/crypto/etc/crypto-configuration.properties]]

[2019-10-15T17:59:38.716+0900] [Payara 4.1] [INFO] [CRYPTO-MSG-1054] [CRYPTO] [tid: _ThreadID=26 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1571129978716] [levelValue: 800] [[
  CRYPTO-MSG-1054: Over-riding properties in STRONGKEYLITE_HOME: ]]

[2019-10-15T17:59:38.717+0900] [Payara 4.1] [INFO] [CRYPTO-MSG-1053] [CRYPTO] [tid: _ThreadID=26 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1571129978717] [levelValue: 800] [[
  CRYPTO-MSG-1053: Configuration log message: Using appliance-configuration.properties from APPLIANCE_HOME directory: /usr/local/strongkey/appliance/etc/appliance-configuration.properties]]

[2019-10-15T17:59:38.717+0900] [Payara 4.1] [INFO] [CRYPTO-MSG-1131] [CRYPTO] [tid: _ThreadID=26 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1571129978717] [levelValue: 800] [[
  CRYPTO-MSG-1131: Over-riding properties in APPLIANCE_HOME:
        appliance.cfg.property.enableddomains.ccspin:
        appliance.cfg.property.replicate: false
        appliance.cfg.property.serverid: 1]]

[2019-10-15T17:59:38.717+0900] [Payara 4.1] [INFO] [CRYPTO-MSG-1053] [CRYPTO] [tid: _ThreadID=26 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1571129978717] [levelValue: 800] [[
  CRYPTO-MSG-1053: Configuration log message: CRYPTO VERSION is: 4.1.0]]

[2019-10-15T17:59:38.822+0900] [Payara 4.1] [WARN] [APPL-ERR-1016] [APPL] [tid: _ThreadID=26 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1571129978822] [levelValue: 900] [[
  APPL-ERR-1016: HMAC Authentication Failed: Expected HMAC: o4c/FX7n10/CrejUIEF9r8l1sKC4qcXh4C/yH1mDZ/8= Produced HMAC: MYsu****************************************]]

[2019-10-15T17:59:38.871+0900] [Payara 4.1] [FATAL] [POC-ERR-5001] [POC] [tid: _ThreadID=34 _ThreadName=http-thread-pool::http-listener-2(10)] [timeMillis: 1571129978871] [levelValue: 1000] [[
  POC-ERR-5001: Error returned while calling the SKFS: Unexpected char 60 at (line no=1, column no=1, offset=0)]]

[2019-10-15T17:59:38.874+0900] [Payara 4.1] [FATAL] [POC-WS-ERR-1000] [POC] [tid: _ThreadID=34 _ThreadName=http-thread-pool::http-listener-2(10)] [timeMillis: 1571129978874] [levelValue: 1000] [[
  POC-WS-ERR-1000: An unexpected error occurred. Please check server logs for details.]]

[2019-10-15T17:59:38.873+0900] [Payara 4.1] [FATAL] [] [] [tid: _ThreadID=34 _ThreadName=http-thread-pool::http-listener-2(10)] [timeMillis: 1571129978873] [levelValue: 1000] [[
  javax.xml.ws.WebServiceException: POC-ERR-5001: Error returned while calling the SKFS: {0}
        at com.strongkey.utilities.Common.parseJsonFromString(Common.java:28)
        at com.strongkey.poc.SKFSClient.getAndVerifySuccessfulResponse(SKFSClient.java:274)
        at com.strongkey.poc.SKFSClient.callServer(SKFSClient.java:262)
        at com.strongkey.poc.SKFSClient.callSKFSRestApi(SKFSClient.java:254)
        at com.strongkey.poc.SKFSClient.preregister(SKFSClient.java:92)
        at com.strongkey.poc.WebauthnService.preregister(WebauthnService.java:137)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
        at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
        at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4836)
        at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:836)
        at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
        at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:64)
        at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
        at sun.reflect.GeneratedMethodAccessor110.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:895)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:835)
        at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
        at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
        at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:895)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:835)
        at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:374)
        at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4808)
        at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4796)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:90)
        at com.sun.proxy.$Proxy486.preregister(Unknown Source)
        at com.strongkey.poc.__EJB31_Generated__WebauthnService__Intf____Bean__.preregister(Unknown Source)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
        at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
        at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
        at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1692)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:258)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:654)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:593)
        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:159)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:483)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:180)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:539)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool]]

[2019-10-15T17:59:38.874+0900] [Payara 4.1] [FATAL] [] [] [tid: _ThreadID=34 _ThreadName=http-thread-pool::http-listener-2(10)] [timeMillis: 1571129978874] [levelValue: 1000] [[
  .java:573)
        at java.lang.Thread.run(Thread.java:748)
]]

**Desktop (please complete the following information):**
 - macOS Mojave 
 - Firefox 69.0.3

**Comment by push2085** [comment]
Hi @Kojiro-Goto ,
This looks like an authentication issue. I see that you did delete the old credentials and recreated new ones.
Can you try the following:
-> update the **_/usr/local/strongkey/poc/etc/poc.properties_** with the new secret key / access key instead of the file you modified and then restart glassfish.
-> Retry registration.

**Comment by Kojiro-Goto** [comment]
Hi @push2085 , 

Thank you for your quick response.

The issue isn't solved yet. Same error.

> Can you try the following:
> -> update the **_/usr/local/strongkey/poc/etc/poc.properties_** with the new secret key / access key instead of the file you modified and then restart glassfish.

I added new secret key and access key into the file you mentioned as follow.

poc.cfg.property.accesskey=**new accesskey**
poc.cfg.property.secretkey=**new secretkey**

and restart glassfish.

Logs are mostly same as I post before.

Do you have any idea for fix it? Do I have to cleanup and re-install software?

**Comment by Kojiro-Goto** [comment]
Hi @push2085 ,

I found the solution of this issue.

The problem is that a system locale is not english.

The server's system locale was LANG=ja_JP.UTF-8.
If the locale isn't US then, maybe, a format of HTTP Date header value is unexpected for hmacRequest().
The values of Date header are:
In case of LANG=ja_JP, the value is "(, 17 10 2019 14:06:35 JST" from FINE log.
In case of LANG=en_US, the value is like "Thu, 17 Oct 2019 14:20:46 JST".

Also I found a description of OpenAPI for challenge, it saids that "The date that the request is being made in the format "EEE, d MMM yyyy HH:mm:ss z"."

Thank you for your time to solve it.
